engine/notifications/notifications.go

```diff
@@ -1,60 +1,69 @@
-// Package notifications is a basic internal notifications module
 package notifications
 
 import (
-	"bytes"
 	"context"
-	"encoding/json"
-	"net/http"
+	"sync"
+
+	"github.com/bitcoin-sv/spv-wallet/models"
 )
 
-// GetWebhookEndpoint will get the configured webhook endpoint
-func (c *Client) GetWebhookEndpoint() string {
-	return c.options.config.webhookEndpoint
+const lengthOfInputChannel = 100
+
+// Notifications - service for sending events to multiple notifiers
+type Notifications struct {
+	inputChannel   chan *models.RawEvent
+	outputChannels *sync.Map //[string, chan *Event]
 }
 
-// Notify will create a new notification event
-func (c *Client) Notify(ctx context.Context, modelType string, eventType EventType,
-	model interface{}, id string) error {
+// AddNotifier - add notifier by key
+func (n *Notifications) AddNotifier(key string, ch chan *models.RawEvent) {
+	n.outputChannels.Store(key, ch)
+}
 
-	if len(c.options.config.webhookEndpoint) == 0 {
-		if c.IsDebug() {
-			c.Logger().Info().Msgf("NOTIFY %s: %s - %v", eventType, id, model)
-		}
-	} else {
-		jsonData, err := json.Marshal(map[string]interface{}{
-			"event_type": eventType,
-			"id":         id,
-			"model":      model,
-			"model_type": modelType,
-		})
-		if err != nil {
-			return err
-		}
+// RemoveNotifier - remove notifier by key
+func (n *Notifications) RemoveNotifier(key string) {
+	n.outputChannels.Delete(key)
+}
 
-		var req *http.Request
-		if req, err = http.NewRequestWithContext(ctx,
-			http.MethodPost,
-			c.options.config.webhookEndpoint,
-			bytes.NewBuffer(jsonData),
-		); err != nil {
-			return err
-		}
+// Notify - send event to all notifiers
+func (n *Notifications) Notify(event *models.RawEvent) {
+	n.inputChannel <- event
+}
 
-		var response *http.Response
-		if response, err = c.options.httpClient.Do(req); err != nil {
-			return err
-		}
-		defer func() {
-			_ = response.Body.Close()
-		}()
-
-		if response.StatusCode != http.StatusOK {
-			// todo queue notification for another try ...
-			c.Logger().Error().Msgf("received invalid response from notification endpoint: %d",
-				response.StatusCode)
+// exchange - exchange events between input and output channels, uses fan-out pattern
+func (n *Notifications) exchange(ctx context.Context) {
+	for {
+		select {
+		case event := <-n.inputChannel:
+			n.outputChannels.Range(func(_, value any) bool {
+				ch := value.(chan *models.RawEvent)
+				n.sendEventToChannel(ch, event)
+				return true
+			})
+		case <-ctx.Done():
+			return
 		}
 	}
+}
+
+// sendEventToChannel - non blocking send event to channel
+func (n *Notifications) sendEventToChannel(ch chan *models.RawEvent, event *models.RawEvent) {
+	select {
+	case ch <- event:
+		// Successfully sent event
+	default:
+		// Channel is full, skip sending event
```

Comment:
maybe it's good idea to log a Warning here

Reply:
I haven't done it because if a notifier input channel were full for some time, many logs would appear with the same content.

But I can add some mechanizm that - based on simple state machine - would produce only one log if the channel becomes full and when it is operable again.

Reply:
Wouldn't `logger.Sample()` with a BurstSample work for that case?